### ISSUE TITLE
font decoding: Th bug

### ISSUE THREAD

**Issue by myrmoteras** [gsautter/goldengate-imagine#16]
@gsautter Th is often wrongly decoded

![image](https://cloud.githubusercontent.com/assets/4609956/21026739/45560418-bd8e-11e6-9416-7ff24519f365.png)
![image](https://cloud.githubusercontent.com/assets/4609956/21026773/70772708-bd8e-11e6-8229-015ba15b1c8a.png)

I will add some more publications when I come along.

[AM52_38_49.pdf](https://github.com/gsautter/goldengate-imagine/files/640538/AM52_38_49.pdf)

**Comment by gsautter** [comment]
Problem with the "Th" ligature is that it doesn't _exist_ in Unicode, so if the to-Unicode mapping provided by the PDF proper is missing or unfriendly, there's no way of getting it, as none of the baseline fonts even has a respective glyph to compare it to.

Baseline fonts are the known fonts (we use the Liberation Fonts) used to compare the rendered glyphs to on decoding, to decide on glyph meaning based on optical similarity. A natural restriction of this approach is that it can only recognize glyphs / characters known to the baseline fonts, i.e., Unicode, and that unfortunately does not apply to this custom ligature.

**Comment by myrmoteras** [comment]
That means we just have to accept an error? "Th" is a bit too common to led it slip, don't you think?

**Comment by gsautter** [comment]
While "Th" is certainly very common, the ligature does not seem to be ... otherwise it would have made it into Unicode, I guess. Here a list of the ligature represented in Unicode: http://en.wikipedia.org/wiki/Typographic_ligature

That does not mean we cannot correct it, it simply means automatically decoding this custom ligature is a really hard one.

**Comment by myrmoteras** [comment]
What is the plan to correct it?

**Comment by gsautter** [comment]
Don't know, change Unicode?

**Comment by gsautter** [comment]
Found a solution: Not sure if you ever tried it, but for a week or so GGI now has had the option to load a PDF with glyph decoding _disabled_ ("PDF Documents (born-digital, render embedded fonts, but do not decode glyphs)"), which incurs whatever Unicode mapping present to simply be used without double-checking. It's nice and fast, but leaves the oddballs as they are (at the risk of having to correct them later).

What can prevent one character from being decoded to its proper Unicode counterpart can prevent another (our notorious out-of-Unicode "Th" ligature) from having its faithful Unicode mapping overwritten with whichever single Unicode character fits best.

The particular mechanism I've added now is that any character that Unicode maps to a known "custom" (i.e., out-of-Unicode, but factually used) ligature gets exempted from further decoding procedures. Works like a charm in case of out "Th", but might well prevent some errors from being corrected in some other cases I can imagine, though have never seen. So be warned: If you some day see some character turn up as a "Th" that actually is none, it's to blame on this very mechanism. Always two sides to this kind of coin ...
